### ISSUE TITLE
Support vector files (KML, SHP, CZML, GeoJSON)

### ISSUE THREAD

**Issue by argallegos** [CesiumGS/cesium-unreal#628]
This is a far-roadmap feature, as implementing it comes with a lot of other considerations, but we should have an issue to track it. 

It's frequently requested on the forums - [01](https://community.cesium.com/t/how-to-use-vector-data-kml-shp/13041), [02](https://community.cesium.com/t/how-to-load-vector-data-like-esris-shp-or-googles-kml-or-geojson/13121), [03](https://community.cesium.com/t/existing-pipelines-to-import-vector-data-is-there-support-for-kml-kmz-shp-files/14147), [04](https://community.cesium.com/t/using-czml-assets-in-unreal-engine/15126/2)

Please feel free to add any other thoughts or potential roadblocks here just so we have it on record.

**Comment by techdragon** [comment]
Just going to point out I’ve repeatedly shelved projects due to this feature being missing. Particularly CZML… Not having support for these formats was and continues to be a huge letdown. I have never wanted to use Cesium in a thatbwasnt in some way dependent on having access to these formats, either as the primary content, or as crucial overlays.

**Comment by shehzan10** [comment]
@techdragon Completely agree with you that CZML and other formats supported in CesiumJS add a ton of functionality to create useful simulations and analytics. We're pushing on all fronts to add this (and more) functionality as it would enable a ton of use-cases, like yours, for the entire community. Its sometimes easy to forget (even for us) that Cesium for Unreal release was less than 6 months, so there is a ton of features we're looking to add to get it close to par with CesiumJS, which has a decade of development behind it.

Please feel free to reach out to me directly if there are specific use-cases we can help with in the short-term while we continue to work on these features.

**Comment by argallegos** [comment]
It looks like a plugin has been created to import this type of data into Unreal before - https://forums.unrealengine.com/t/plugin-import-gis-data-into-ue4/107459, but it doesn't seem maintained currently. Might be helpful as we look into our own implementation.

**Comment by techdragon** [comment]
I'd like to see specific issues for each of these formats, with the underlying work noted as blockers. Its pretty obvious supporting all the various vector formats is an "epic" task, and its all too easy for interested parties on the outside, like myself and others to see very little traction on these kinds of tasks, since naturally a lot of work needs to be done before such big tickets are marked as "shipped". 

Also, that lets people express their support for each format and allow some kind of prioritisation to be performed. For instance CZML and GeoJSON share the need for JSON parsing functionality, KML needs an XML parser, and SHP needs a custom parser for the ESRI Shapefile format. (This also happens to be my personal prioritisation order of these four formats) All the formats will likely need a vector rendering layer to handle common vector rendering coordinate transforms and such, and would probably need a number of renderable object types they can use to show the various data types in the world such as marker, line, polygon, etc types and these should obviously be extensible so that we can implement more advanced rendering of these primitives if needed (for example, rendering markers that vary in size based on data from the GeoJSON). While its unfair to demand that this be a top priority, its a lot easier for us the community to plug gaps/help out/ build our own... if we  _see_ the incremental progress on this feature and know that fundamental components are built so we can give the individual functionality parts a shot. 

Personally I'd feel much more confident taking a shot at building my own marker/point (and maybe even lines) rendering code if I had the parsing & world coordinate transform taken care of by people that know how all that works a lot better than me. Because while my point could be just enough code to do what I need, its useless to me if I can't keep my points in the right place when the existing raster data map layers move and zooms. I could probably do a decent enough job on a file parser too to be honest. So having paid employee time dedicated to a foundational part necessary for this functionality could be very valueable (the vector rendering layer to handle common vector rendering coordinate transforms I mentioned earlier)   

That said... as I write all this out... and read through the https://github.com/CesiumGS/cesium-native code to try and see how the raster layers work since I was looking for something to suggest "we just need a vector version of this class", it feels like a lot of the the Unreal Engine specific magic is in the `CesiumGeoreference*` files and it might not be quite as much work as I first thought to implement a basic "vector point" but as I don't have the time right now to dig in deeper, I can't say for sure how hard it would be to also add lines or polygons. Lack of a C++ CZML parser isn't a huge deal as its just a specific JSON format so something shouldn't be too hard to cobble together for at least testing/prototyping purposes.

**Comment by davesauce14** [comment]
Hi team, is there any progress on this? I see this is from over a year ago

**Comment by IndigoCloud6** [comment]
Hi team, is there any progress on this?

**Comment by Neurowerk** [comment]
Hi, team, this is will be significant as well, to support CZML files for Cesium for Unreal.

**Comment by lzp0070007** [comment]
Hi, team, this is will be significant as well, to support geojson files for Cesium for Unreal.

**Comment by RayDog1995** [comment]
Trying to simplify flight crash recreation animations, attempted fs2020 and ran into issues, played around with cesium and if it would allow a kml file to be imported and used as a spline or something similar so I can easily add a plane model to have it follow the recorded path captured from the blackbox. Also id hope altitude data would be recognized and speed between data points.

**Comment by kring** [comment]
@RayDog1995 have you seen this "Build a Flight Tracker" tutorial?
https://cesium.com/learn/unreal/unreal-flight-tracker/

**Comment by RayDog1995** [comment]
Yes, several times. That tutorial uses precompiled .csv files, have you seen this thread where everyone is asking for kml support in cesium plugin for unreal just like how cesiumjs can use kml files: https://community.cesium.com/t/how-to-use-vector-data-kml-shp/13041/2

**Comment by kring** [comment]
Yes, I understand that people want KML support. But that's a large feature, and we currently have more pressing priorities so we won't be able to start it right away. So I'm trying to find you a workaround in the meantime. If you're trying to make a plane follow a path, using a CSV file like in the flight tracker tutorial seems like it could be a viable option. Even if you have to convert the source data to CSV first.

**Comment by L0CKZ0R** [comment]
Bumping for update. The ability to bring in KML data like a part would be amazing. The CSV trick is very hacky and so far I haven't got it to work.

**Comment by azrogers** [comment]
Another request for GeoJSON from the forums: https://community.cesium.com/t/geojson-from-cesium-ion-to-unreal-engine/31495

**Comment by azrogers** [comment]
Apropos of nothing, I had an idea for how we might implement this across native runtimes using [msdfgen](https://github.com/Chlumsky/msdfgen), which is a library for generating Multi-Channel Signed Distance Fields. This means, when loading a vector, we'd load our data into a `msdfgen::Shape` (using code that would probably look a lot like its [SVG implementation](https://github.com/Chlumsky/msdfgen/blob/755c45b4445baa12bfead90fa7d4d2259f2a14c4/ext/import-svg.cpp)) and render it out to a texture, which would look something like:

```cpp
Shape shape;
if (loadKmlFile(shape, filename)) {
	Bitmap<float, 3> msdf(64, 64);
	SDFTransformation t(...);
	generateMSDF(msdf, shape, t);
}
```

Then we copy that bitmap to a texture in whatever runtime we're using and treat it like a raster overlay, with the following shader code:
```glsl
in vec2 texCoord;
out vec4 color;
uniform sampler2D msdf;
uniform vec4 bgColor;
uniform vec4 fgColor;
uniform float pxRange;

float median(float r, float g, float b) {
    return max(min(r, g), min(max(r, g), b));
}

float screenPxRange() {
    vec2 unitRange = vec2(pxRange)/vec2(textureSize(msdf, 0));
    vec2 screenTexSize = vec2(1.0)/fwidth(texCoord);
    return max(0.5*dot(unitRange, screenTexSize), 1.0);
}

void main() {
    vec3 msd = texture(msdf, texCoord).rgb;
    float sd = median(msd.r, msd.g, msd.b);
    float screenPxDistance = screenPxRange()*(sd - 0.5);
    float opacity = clamp(screenPxDistance + 0.5, 0.0, 1.0);
    color = mix(bgColor, fgColor, opacity);
}
```

This should give us very clean, crisp vector rendering on the GPU. Of course, this is oversimplifying - the `loadKmlFile` function (or `loadShpFile` or `loadCzmlFile` or whatever) is quite a task, so it's a bit of a "draw the rest of the owl" situation. But I think this is an approach that would let us handle the vast majority of the implementation in cesium-native with minimal work for implementation in each runtime - and without rasterizing the data.

**Comment by azrogers** [comment]
Hah. I wanted to see if there were any techniques for multi-color SDFs, and it looks like Adobe published a technique... in [US patent number #10403005](https://patents.justia.com/patent/10403005). Really cool technique, using the four color theorem to divide up the image into four separate color planes, each of regions of the original color without any adjacent pixels that might cause bleeding. The color planes are dilated to make sure each shape will be completely covered by its corresponding plane. Then in the shader you select the best SDF for the current point and use the color from its color plane. Would use eight or ten textures instead of one, but it could work. Shame it's patented. Maybe we can use this technique in whatever Cesium for Unreal version comes out in 2039.

We could [build a texture atlas of all of the 24 RGB bit planes](https://gpuhacks.wordpress.com/2013/07/08/signed-distance-field-rendering-of-color-bit-planes/). If we used Valve's two-channel SDF approach instead of the three-channel msdfgen approach, we could pack a full-color 128x128 SDF into a 512x512 RGBA texture. But this feels more like a science project than a realistic proposal.

**Comment by azrogers** [comment]
Did not mean to close the issue :sob:

**Comment by j9liu** [comment]
Came up again on the [forum](https://community.cesium.com/t/geojson-unreal/37413).

**Comment by azrogers** [comment]
As this has continued to come up in our own discussions as well as in the community, I thought I might take a moment to list out my current thoughts on the matter before things speed up again after the new year and leave me without the time.

It seems to me like there's two parts to implementing vector support in Cesium Native: parsing and rendering. From comments in this thread and on the forums, it seems like we don't necessarily have to implement rendering for parsing to be useful. The ability to obtain a path or a list of points from a vector file would be a useful feature on its own. 

For parsing, we can generalize across the formats: we need to support the idea of geometry *objects* (points, lines, polygons, etc), and geometry *features* (a geometry object annotated with some kind of data, such as a placemark). A basic implementation of KML, SHP, CZML, and GeoJSON files could be accomplished by supporting the following geometry objects:
- A set of X, Y, Z cartesian coordinates, referred to as a Point (KML, CZML, GeoJSON, SHP)
- A set of connected line segments, referred to as a Polyline (CZML, SHP) or a LineString (KML, GeoJson)
    - KML LinearRing types could be read as this type instead.
- A connected set of three or more vertices, referred to as a Polygon (KML, CZML, GeoJSON, SHP)
- A reference to a supported 3D model format, referred to as Model (CZML, KML)
- A collection of one or more geometry objects, referred to as GeometryCollection (GeoJSON) or MultiGeometry (KML)
    - This would also support SHP & GeoJson's `MultiPoint` types and GeoJSON's `MultiLineString` and `MultiPolygon` types
  
Features are essentially a reference to a geometry object along with some amount of arbitrary data. For KML, this is a name, description, type-specific data, and custom metadata. For GeoJSON, this is an arbitrary JSON object. We could treat features as bundles of arbitrary data without a schema for the moment, and save any implementation of specific features for when they're needed.

This short description here would allow us to support GeoJSON and SHP files almost entirely. Many KML features like styles, views, and tours would be unsupported with this subset, but it would still support most of the actual geospatial data in a KML file. CZML is a beast that I don't even want to think about implementing at the moment.

For the moment, I think it would be worth implementing a GeoJSON parser. We already have significantly more code for working with JSON data than XML data, so I think it's a natural fit, and the spec is short and focused enough that we don't need a lot of work to implement it 100%. It would also have the side effect of helping us implement geocoding. The [Pelias](https://pelias.io/) geocoder returns GeoJSON, as does the Cesium ion geocoder (the CesiumJS implementation of the ion geocoder is actually just a facade around the `PeliasGeocoderService`!).

**Comment by lilleyse** [comment]
I think the big question is whether vector formats would transcode to glTF (plus extensions) or some other format.

Some discussion about polylines in glTF here: https://github.com/iTwin/itwinjs-core/issues/7311.

**Comment by kring** [comment]
I've long been of the opinion that "parsing is easy" and what's challenging about vector support is 1) rendering, and 2) level-of-detail. Even if that's true, that doesn't necessarily mean that parsing is useless.

Before we start implementing anything, though, I think we should be clear about what capabilities we're providing and how they will integrate into Unreal and our other game engines. For example, maybe GeoJSON points/lines/polygons could feed into Unreal's [Procedural Content Generation](https://dev.epicgames.com/documentation/en-us/unreal-engine/procedural-content-generation-overview) system? I say that without entirely know what I mean by it! :)

Just as an example of what's possible here, my old team at Terria got a ton of mileage out of a relatively simple solution that imports GeoJSON and other vector sources, tiles and simplifies the features in-browser using [geojson-vt](https://github.com/mapbox/geojson-vt), and then symbolizes/styles and renders them to images using [protomaps](https://www.npmjs.com/package/protomaps-leaflet). The images are then draped over terrain as raster overlays.

**Comment by techdragon** [comment]
Thanks for taking the time to elaborate @azrogers ... and it validates most of what i wrote more than 3 years ago here https://github.com/CesiumGS/cesium-unreal/issues/628#issuecomment-927258693
> Also, that lets people express their support for each format and allow some kind of prioritisation to be performed. For instance CZML and GeoJSON share the need for JSON parsing functionality, KML needs an XML parser, and SHP needs a custom parser for the ESRI Shapefile format. (This also happens to be my personal prioritisation order of these four formats) All the formats will likely need a vector rendering layer to handle common vector rendering coordinate transforms and such, and would probably need a number of renderable object types they can use to show the various data types in the world such as marker, line, polygon, etc types and these should obviously be extensible so that we can implement more advanced rendering of these primitives if needed (for example, rendering markers that vary in size based on data from the GeoJSON). While its unfair to demand that this be a top priority, its a lot easier for us the community to plug gaps/help out/ build our own... if we _see_ the incremental progress on this feature and know that fundamental components are built so we can give the individual functionality parts a shot.

While I'm a developer myself and get the difficulty of prioritising things on such large projects, I do find it completely baffling that we still don't even have a vector graphics primitive that integrates correctly with camera so that we could try and reliably build our own data renders on top of the world imagery.

I continue to be utterly baffled that the more powerful framework that Unreal provides for visualising lots of 3D data, is still not able to make full use of the engine. Rather than abandoning a personal project that I would have liked to use Cesium-Unreal for... I ended up building it with CesiumJS and stuck a web browser inside Unreal engine... that's a pretty large and ugly work around for something that at least I would have been considered essential functionality that belonged in the "make it work in the first 1 or 2 years at the most" priority... not consigned to the backlog for nearly 4 years now with no clear sign it will ever be addressed.

**Comment by kring** [comment]
Hi @techdragon,
I appreciate your frustration. I wish this were done already, too. But you have some options:
1. Implement it yourself. Everything is open source, and we welcome pull requests.
2. Implement a limited workaround for your particular vector data use-case. We've offered advice on this on the community forum in the past. For example, Unreal's built-in JSON-parsing capabilities can handle GeoJSON, and from there you can write C++ or Blueprint to create Unreal objects from the data.
3. Fund the development of this. Cesium doesn't typically do work-for-hire, but when work is closely aligned with our roadmap (like this is), a funded contract can shift priorities by getting this particular work done sooner or focusing the efforts on the parts of it that are most immediately impactful for you.
4. Wait until we do it, among the very many other things competing for our development time.

**Comment by darcyvdd** [comment]
This has come up again on the [forum](https://community.cesium.com/t/visualizing-more-metadata/45799)

**Comment by azrogers** [comment]
After substantial work on the vector data front, I think it's time to close this issue. Here's my reasoning:

1. We now have support for vector data through the `CesiumGeoJsonDocument`, `CesiumGeoJsonDocumentRasterOverlay`, and now `CesiumVectorTilesRasterOverlay` classes. We also now have the ability to load GeoJSON as glTF thanks to CesiumGS/cesium-native#1334. 
2. KML and SHP files can be easily re-encoded as GeoJSON documents through any number of off-the-shelf tools. This also goes for any number of other formats not mentioned here. It would be better to use these conversion tools, which have gone through substantial amounts of effort to properly support all of these formats, than to implement support for these ourselves. In the worst-case scenario, a user could bring in libkml or ShapefileCpp as a dependency to load these file types and encode their data in a `GeoJsonDocument` for use with the Cesium for Unreal vector data tools.
3.  CZML is a legacy format. It has not seen any attention in a decade. In general, the way Cesium handles things these days is to try to work with the broadest, most compatible open formats rather than design our own. CZML pre-dates 3D Tiles, it even pre-dates glTF 2.0.
4. Besides that, CZML is a very unique format. I wouldn't really call it a vector format. It's more like a time-series scene format. We can't transcode it to GeoJSON or glTF or 3D Tiles or any other format without losing a substantial amount of functionality and information. It would be an entirely bespoke implementation that would require wide-ranging changes in the way Cesium for Unreal works, to be able to support a legacy format that is seeing very little use these days.

This is very much not to say that we are done with all the vector format support we are ever going to add to Cesium for Unreal. We have a number of things planned for GeoJSON and vector tiles in the coming months and years. This also does not close the book on supporting any of the above formats. But I think that this broad "support vector files" ticket here has outlived its usefulness. 

If there are formats or features for vector data that people want to see brought to Cesium for Unreal, I would encourage people to open new issues in this repo or in CesiumGS/cesium-native for those specific tasks.
